small/ibuf.c

```diff
@@ -61,6 +61,25 @@ ibuf_reinit(struct ibuf *ibuf)
 	ibuf_create(ibuf, slabc, start_capacity);
 }
 
+/** A helper function for ibuf_reserve_slow() and ibuf_shrink(). */
+static void *
+ibuf_resize(struct ibuf *ibuf, size_t new_capacity)
+{
+	size_t used = ibuf_used(ibuf);
+	assert(new_capacity >= used);
+	struct slab *slab = slab_get(ibuf->slabc, new_capacity);
+	if (slab == NULL)
+		return NULL;
+	char *ptr = (char *) slab_data(slab);
+	memcpy(ptr, ibuf->rpos, used);
+	ibuf_destroy(ibuf);
```

Comment:
I think we should not use `ibuf_destroy` here. `ibuf_destroy` is expected to destroy entire buffer while we need to just destroy the slab.
For example in `ibuf_destroy` we can trash all the contents of `ibuf`, like we usually do. Though the most of the members will be initialized below, `slabc` and `start_capacity` will be corrupted and that will lead to a hard debugged errors..
Generally `_destroy` must do exactly what is expected (destruction) and should be called only if exactly destruction is needed.

Reply:
Also, I find `ibuf_shrink` so trivial that I wouldn't factor out the common part of `ibuf_reserve` and `ibuf_shrink` into a separate helper function. A few duplicated lines should be okay as long as the code is readable.

Reply:
Done.